starfish/codebook.py

```diff
@@ -356,13 +357,101 @@ def to_json(self, filename: str) -> None:
+    @staticmethod
+    def _normalize_features(array: Union["Codebook", IntensityTable], norm) \
+            -> Tuple[Union["Codebook", IntensityTable], np.ndarray]:
+        """unit normalize each feature of array
+
+        Parameters
+        ----------
+        array : Union[Codebook, IntensityTable]
+            codebook or intensity table containing (ch, r) features to normalize
+        norm : int
+            the norm to apply to each feature
+
+        See Also
+        --------
+        The available norms for this function can be found at the following link:
+        https://docs.scipy.org/doc/numpy-1.14.0/reference/generated/numpy.linalg.norm.html
+
+        Returns
+        -------
+        Union[IntensityTable, Codebook] :
+            IntensityTable or Codebook containing normalized features
+        np.ndarray :
+            A 1 dimensional numpy array containing the feature norms
+
+        """
+        feature_traces = array.stack(traces=(Indices.CH.value, Indices.HYB.value))
+        norm = np.linalg.norm(feature_traces.values, ord=norm, axis=1)
+        array = array / norm[:, None, None]
+
+        # if a feature is all zero, the information should be spread across the channel
+        partitioned_intensity = 1 / np.product(array.shape[1:])
+        array = array.fillna(partitioned_intensity)
+
+        return array, norm
+
+    @staticmethod
+    def _approximate_nearest_code(
+            norm_codes: "Codebook", norm_intensities: xr.DataArray
+    ) -> Tuple[np.ndarray, np.ndarray]:
+        """find the nearest code for each feature using the ball_tree approximate NN algorithm
+
+        Parameters
+        ----------
+        norm_codes : Codebook
+            codebook with each code normalized to unit length (sum = 1)
+        norm_intensities : IntensityTable
+            intensity table with each feature normalized to unit length (sum = 1)
+
+        Returns
+        -------
+        np.ndarray : qualities
+            the quality of each feature's match to its closest code (equal to 1 - distance)
+        np.ndarray : gene_ids
+            the gene that corresponds to each matched code
+
+        """
+        # TODO ambrosejcarr: see if features can be pre-masked to reduce NN calculations
+        linear_codes = norm_codes.values.reshape(norm_codes.shape[0], -1)
+        linear_features = norm_intensities.values.reshape(norm_intensities.shape[0], -1)
+        # reshape into traces
+        nn = NearestNeighbors(n_neighbors=1, algorithm='ball_tree').fit(linear_codes)
+        distances, indices = nn.kneighbors(linear_features)
+        gene_ids = np.ravel(norm_codes.indexes[Codebook.Constants.GENE.value].values[indices])
+        qualities = 1 - np.ravel(distances)
+
+        return qualities, gene_ids
+
+    def decode_euclidean(
+            self, intensities: IntensityTable, max_distance: Number, min_intensity: Number=0,
+            norm: int=2, return_norms: bool=False,
```

Comment:
we should move away from exposing defaults -- methods are too variable for now to expose meaningful defaults imo